java/fury-core/src/main/java/org/apache/fury/Fury.java

```diff
@@ -1221,15 +1240,55 @@ public Object deserializeJavaObjectAndClass(FuryReadableChannel channel) {
     return deserializeJavaObjectAndClass(buf);
   }
 
+  @SuppressWarnings("unchecked")
   @Override
   public <T> T copy(T obj) {
     if (Objects.isNull(obj)) {
       return null;
     }
-    if (obj instanceof FuryCopyable) {
-      return (T) ((FuryCopyable) obj).copy(this);
-    } else {
-      return (T) classResolver.getOrUpdateClassInfo(obj.getClass()).getSerializer().copy(obj);
+    try {
+      copyDepth++;
+      if (Objects.isNull(originToCopyMap)) {
+        originToCopyMap = new IdentityMap();
+      }
+      Object newObj = originToCopyMap.get(obj);
+      if (Objects.nonNull(newObj)) {
+        return (T) newObj;
+      }
+      if (copyRefTracking) {
+        originObjRef = obj;
+      }
+      T copy;
+      if (obj instanceof FuryCopyable) {
```

Comment:
this check is expensive, we may better put it into a Serializer#copy

Reply:
We'd better to create a new Serializer like FuryCopyableSerializer for this instead of using isntanceof here. It's very slow.

The code could looks like in getSerializer:
```java
serializer = XXX;
if (FuryCopyable.class.isAssignableFrom(cls)) {
  serializer = new FuryCopyableSerializer(serializer);
}

class FuryCopyableSerializer {
 private Serializer serializer;

write(...) {
serializer.write(...);
}

read(...) {
serializer.read(...);
}

T copy(T x) {
        return (T) ((FuryCopyable<?>) obj).copy(x);
}

}
```

Reply:
This looks like changing `instanceof` to `Class#isAssignableFrom`, maybe we can do this:
```java
Class<?> cls = obj.getClass();
if (FuryCopyable.class.isAssignableFrom(cls)) {
    return (T) ((FuryCopyable<?>) obj).copy(this);
} else {
    return (T) classResolver.getOrUpdateClassInfo(cls).getSerializer().copy(obj);
}
```

The benefits of doing this are:
1. No need to instantiate an additional `FuryCopyableSerializer` object
2. For `write/read`, do not call `FuryCopyableSerializer`, but use the original `Serializer`